File: src/compose.ts
/**
 * Turns a capture into the text fields GlassFrog will store.
 *
 * Revises KTD5 in one respect, forced by live API behaviour rather than by
 * preference — see the note on the headline below. The decision's *substance*
 * is preserved: the provenance marker leads its field and is never truncated,
 * so nothing the practitioner captured can silently destroy it.
 *
 * This module is pure. It is the one place R11 can fail silently — a filed item
 * missing its marker looks perfectly fine in GlassFrog while quietly making the
 * triage-survival metric uncomputable.
 */
import type { Capture, PageContext } from './types.ts';

/**
 * Stable by contract. Triage-survival matching and the read-back reconciliation
 * contemplated in issue #9 both key off this exact string; changing it orphans
 * every item filed before the change.
 */
export const PROVENANCE_MARKER = '[glassfrog-clipper]';

/** R7's cap, applied to each page-derived evidence field on its own. */
export const EVIDENCE_FIELD_LIMIT = 4000;

/**
 * A headline is a headline. 200 is GlassFrog's own cap on a tension `label`,
 * adopted uniformly so an action's `description` cannot become a wall of text
 * either. Verified against the live API: the field rejects more.
 */
export const HEADLINE_LIMIT = 200;

const ELLIPSIS = '…';

/**
 * Code-point-safe truncation. Slicing by UTF-16 code unit can split a surrogate
 * pair and leave a lone half in the filed item, so this counts characters the
 * way a reader would.
 */
export function truncate(text: string, limit: number = EVIDENCE_FIELD_LIMIT): string {
  const points = Array.from(text);
  if (points.length <= limit) return text;
  return points.slice(0, Math.max(0, limit - 1)).join('') + ELLIPSIS;
}

/**
 * Removes the URL's `userinfo` component — `user:password@`, and the bare
 * `token@` and `:password@` forms — from a captured URL.
 *
 * R7 carries the URL as evidence, and a magic-link, password-reset, or
 * signed-URL page turns a zero-decision keystroke into an export of a
 * credential the practitioner never chose to share. Userinfo is the one part of
 * a URL that is *definitionally* a credential, so removing it costs no
 * decision — which is what keeps this on the right side of STRATEGY.md's resist
 * test. Secrets carried in the query string or fragment are a judgement call,
 * not a component, and R7 carries those as-is by design.
 *
 * A URL with no userinfo is returned byte-identical rather than round-tripped
 * through `URL`. Serialising every capture would lowercase hosts, add trailing
 * slashes, and re-encode paths — rewriting the evidence on the overwhelming
 * majority of captures that were never at risk.
 *
 * An unparseable string is returned unchanged, and that fallback is *unsafe in
 * the abstract*: `new URL` throws on `https://alice:hunter2@example.test:99999/`
 * and on `//alice:hunter2@example.test/`, each of which still spells out a
 * password. It is safe here only because the input is `chrome.tabs.Tab.url` —
 * Chrome's own parser rejects those same strings before a tab can commit to one,
 * so no such value ever arrives. Reusing this function on a URL from anywhere
 * else breaks that argument, which is why the Swift port for the share sheet
 * does not inherit it.
 */
export function stripUrlCredentials(url: string): string {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return url;
  }

  if (!parsed.username && !parsed.password) return stripNestedUrlCredentials(url);

  parsed.username = '';
  parsed.password = '';
  return parsed.href;
}

/**
 * The schemes that carry a whole URL inside their own path.
 *
 * `new URL('view-source:https://alice:hunter2@example.test/')` parses happily
 * and reports an empty `username` and `password`, because the inner address is
 * just path text to it — so the plain check above waves the credential straight
 * through. That is the exact shape this guard exists to stop: nothing fails, and
 * the password is filed.
 *
 * Recursing rather than reaching for a lexical `//...@` strip is deliberate. A
 * pattern loose enough to catch these also matches a query string like
 * `?next=//user:pass@host`, and rewriting that would destroy evidence in a URL
 * that was never a credential. Recursion terminates because each level removes
 * its scheme prefix, so the string is strictly shorter every time.
 */
const NESTED_SCHEME = /^(view-source|blob|filesystem):(.+)$/is;

function stripNestedUrlCredentials(url: string): string {
  const [, scheme, rest] = NESTED_SCHEME.exec(url) ?? [];
  if (!scheme || !rest) return url;

  const inner = stripUrlCredentials(rest);
  // Byte-identical unless the inner URL actually changed, so a `blob:` URL that
  // never held a credential is returned exactly as the browser reported it.
  return inner === rest ? url : `${scheme}:${inner}`;
}

export type Composed =
  | { kind: 'tension'; body: string }
  | { kind: 'action'; description: string; note: string }
  /**
   * Only a project carries `link`. `ActionInput` has no such field and neither
   * does a tension — verified against the SDK's own types — so there is nothing
   * to invent an equivalent for on those two paths.
   */
  | { kind: 'project'; description: string; note: string; link?: string };

/**
 * Marker first, then as much of the title as fits inside HEADLINE_LIMIT.
 *
 * The marker leads and is never truncated, so R11 holds no matter how long the
 * title is — the property KTD5 was written to guarantee.
 */
export function headline(page: PageContext): string {
  const budget = HEADLINE_LIMIT - PROVENANCE_MARKER.length - 1;
  const title = truncate(page.title.trim(), budget).trim();
  return title ? `${PROVENANCE_MARKER} ${title}` : PROVENANCE_MARKER;
}

/**
 * The practitioner's own words come before the evidence the machine gathered —
 * whoever reads this in triage is looking for the thought, not the URL.
 */
function detail(capture: Capture): string {
  const parts: string[] = [];

  const note = capture.note?.trim();
  if (note) parts.push(note);

  const evidence: string[] = [];
  const url = truncate(capture.page.url).trim();
  if (url) evidence.push(url);

  const selection = capture.page.selection?.trim();
  if (selection) evidence.push(truncate(selection));

  if (evidence.length > 0) parts.push(evidence.join('\n\n'));

  return parts.join('\n\n');
}

export function compose(capture: Capture): Composed {
  const head = headline(capture.page);
  const body = detail(capture);

  // R4 / KD2: an unset work type is a tension. No `status` is composed for one —
  // v5 auto-computes unprocessed/processed from associations and accepts only
  // `archived` from a client, verified live.
  switch (capture.workType) {
    case 'action':
      return { kind: 'action', description: head, note: body };
    case 'project':
      // The URL goes to `link` AND stays in the note. The note is the
      // human-readable evidence block and is subject to R7's truncation; `link`
      // is the single canonical field GlassFrog renders the project as linked
      // from, and truncation must never reach it. Both, deliberately.
      //
      // An empty URL is omitted rather than sent blank: `link: ''` would read
      // as a link that exists and is broken.
      return {
        kind: 'project',
        description: head,
        note: body,
        ...(capture.page.url.trim() ? { link: capture.page.url } : {}),
      };
    default:
      // No `label`. The generated OpenAPI types list it on TensionInput, but
      // the API rejects it on create — the tension is created first and the
      // label PATCHed separately, which this path deliberately does not do.
      //
      // A second call would be a second failure point between the POST and the
      // marker landing, and KTD7's at-most-once turns on there being exactly
      // one write per capture. A capture whose marker went missing because the
      // worker died mid-PATCH is precisely the silent R11 failure this module
      // exists to prevent. The marker leads the body instead.
      return { kind: 'tension', body: [head, body].filter(Boolean).join('\n\n') };
  }
}

File: apple/GlassFrogClipperCore/Sources/GlassFrogClipperCore/Compose.swift
import Foundation

/// Turns a capture into the text fields GlassFrog will store.
///
/// A deliberate, line-for-line port of `src/compose.ts`. It exists twice because
/// the Share Extension never runs the web extension — it is handed a URL by the
/// share sheet and files natively — so the composition has to happen in Swift
/// too.
///
/// Two implementations of one function is a liability, and the reason it is
/// accepted here rather than routed around is that the alternative is worse: the
/// Share Extension would otherwise have to wake the web extension to compose for
/// it, which is a second failure point on the capture path for no gain. The
/// liability is contained by `test/fixtures/compose-cases.json`, a golden file
/// both languages assert against. Neither can drift without a visible diff.
///
/// **If you change anything in this file, regenerate that fixture from the
/// TypeScript side and make both suites pass.** ADR 0004 makes the marker the
/// basis of the triage-survival metric, and an item filed with a marker one
/// character adrift is invisible to that metric while looking perfectly ordinary
/// in GlassFrog.
public enum Compose {

    /// Stable by contract. Changing it orphans every item filed before the
    /// change. Must equal `PROVENANCE_MARKER` in `src/compose.ts`.
    public static let provenanceMarker = "[glassfrog-clipper]"

    /// R7's cap, applied to each page-derived evidence field on its own.
    public static let evidenceFieldLimit = 4000

    /// GlassFrog's own cap on a tension `label`, adopted uniformly so an
    /// action's `description` cannot become a wall of text either.
    public static let headlineLimit = 200

    private static let ellipsis = "…"

    /// Exactly what JavaScript's `String.prototype.trim` removes.
    ///
    /// Not `.whitespacesAndNewlines`, which differs at both ends: it includes
    /// U+0085 (NEL), which JavaScript does not trim, and omits U+FEFF, which
    /// JavaScript does. Either difference produces a body that survives every
    /// ASCII test and diverges from the TypeScript output on real-world text
    /// pasted out of a word processor.
    private static let jsWhitespace: CharacterSet = {
        var set = CharacterSet.whitespaces          // Unicode Zs, plus TAB
        set.formUnion(CharacterSet(charactersIn: "\u{000A}\u{000B}\u{000C}\u{000D}"))
        set.insert(charactersIn: "\u{FEFF}")        // trimmed by JS, not by Foundation
        set.remove(charactersIn: "\u{0085}")        // trimmed by Foundation, not by JS
        return set
    }()

    private static func trimmed(_ text: String) -> String {
        text.trimmingCharacters(in: jsWhitespace)
    }

    /// Truncation counted in Unicode scalars, matching `Array.from(text)`.
    ///
    /// This is the single most likely place for the two implementations to
    /// diverge, and the divergence is silent. Swift's `String` is a collection
    /// of *grapheme clusters*, so the obvious `text.prefix(limit)` counts
    /// user-perceived characters. For the fixture's family-emoji title that is
    /// 47 where JavaScript counts 200 — a wildly different headline, from code
    /// that reads correctly and passes every ASCII case.
    public static func truncate(_ text: String, limit: Int = evidenceFieldLimit) -> String {
        let scalars = Array(text.unicodeScalars)
        guard scalars.count > limit else { return text }
        let kept = scalars.prefix(max(0, limit - 1))
        var view = String.UnicodeScalarView()
        view.append(contentsOf: kept)
        return String(view) + ellipsis
    }

    /// Marker first, then as much of the title as fits inside `headlineLimit`.
    ///
    /// The marker leads and is never truncated, so R11 holds no matter how long
    /// the title is.
    public static func headline(_ page: PageContext) -> String {
        let budget = headlineLimit - provenanceMarker.unicodeScalars.count - 1
        let title = trimmed(truncate(trimmed(page.title), limit: budget))
        return title.isEmpty ? provenanceMarker : "\(provenanceMarker) \(title)"
    }

    /// The practitioner's own words come before the evidence the machine
    /// gathered — whoever reads this in triage is looking for the thought, not
    /// the URL.
    static func detail(_ capture: Capture) -> String {
        var parts: [String] = []

        let note = trimmed(capture.note ?? "")
        if !note.isEmpty { parts.append(note) }

        var evidence: [String] = []
        let url = trimmed(truncate(capture.page.url))
        if !url.isEmpty { evidence.append(url) }

        let selection = trimmed(capture.page.selection ?? "")
        if !selection.isEmpty { evidence.append(truncate(selection)) }

        if !evidence.isEmpty { parts.append(evidence.joined(separator: "\n\n")) }

        return parts.joined(separator: "\n\n")
    }

    /// What GlassFrog is actually sent.
    ///
    /// R4 / KD2: an unset work type is a tension, and no `status` is composed
    /// for one.
    public static func compose(_ capture: Capture) -> Composed {
        let head = headline(capture.page)
        let body = detail(capture)

        switch capture.workType {
        case .action:
            return .action(description: head, note: body)
        case .project:
            // The URL goes to `link` AND stays in the note. The note is the
            // human-readable evidence block and is subject to R7's truncation;
            // `link` is the single canonical field GlassFrog renders the project
            // as linked from, and truncation must never reach it. Both,
            // deliberately.
            //
            // An empty URL is omitted rather than sent blank: `link: ""` would
            // read as a link that exists and is broken.
            let url = capture.page.url.trimmingCharacters(in: jsWhitespace)
            return .project(description: head, note: body, link: url.isEmpty ? nil : capture.page.url)
        case .tension, .none:
            // No `label`. The generated OpenAPI types list it on TensionInput,
            // but the API rejects it on create — see ADR 0004. The marker rides
            // at the head of the body instead, so there is exactly one write per
            // capture and no window in which the marker can go missing.
            return .tension(body: [head, body].filter { !$0.isEmpty }.joined(separator: "\n\n"))
        }
    }
}

/// The composed result, mirroring the `Composed` union in `src/compose.ts`.
public enum Composed: Equatable, Sendable {
    case tension(body: String)
    case action(description: String, note: String)
    /// Only a project carries `link`. `ActionInput` has no such field and
    /// neither does a tension, so there is nothing to invent an equivalent for
    /// on those two paths.
    case project(description: String, note: String, link: String?)

    /// The field the provenance marker leads, whichever shape this is. R11 is
    /// stated over this rather than over each case, so a new case cannot quietly
    /// escape the check.
    public var markedField: String {
        switch self {
        case let .tension(body): return body
        case let .action(description, _): return description
        case let .project(description, _, _): return description
        }
    }
}
